### PR TITLE
Add Eigen3 support

### LINKED ISSUE #12: Replace gsl in linalg with LAPACK calls

### PR DESCRIPTION
**Pull request by junghans**
Part of votca/votca#73

Fix #12

### CODE REVIEW

File: include/votca/tools/spline.h [hunk 1]
```diff
@@ -47,15 +44,15 @@ class Spline
...
     /**
      * \brief Fit spline through noisy (x,y) values. Points on resulting fitted spline can be obtained via Calculate().
      * \param x values of data to be fitted
      * \param y values of data to be fitted
      * both vectors must be of same size
      */
-    virtual void Fit(ub::vector<double> &x, ub::vector<double> &y) = 0;
+    virtual void Fit(Eigen::VectorXd &x, Eigen::VectorXd &y) = 0;
 
     /**
```
Comment (JoshuaSBrown) [inline include/votca/tools/spline.h]: Can you make any of these const?

File: include/votca/tools/spline.h [hunk 2]
```diff
@@ -152,41 +149,41 @@ class Spline
...
-    ub::vector<double> _f2;
+    Eigen::VectorXd _f2;
 };
 
 template<typename vector_type1, typename vector_type2>
 inline void Spline::Calculate(vector_type1 &x, vector_type2 &y)
 {
-    for(size_t i=0; i<x.size(); ++i)
+    for(int i=0; i<x.size(); ++i)
         y(i) = Calculate(x(i));
```
Comment (JoshuaSBrown) [inline include/votca/tools/spline.h]: Again I think you will run into problems with this if you are comparing an int with a size_t    you may need to cast the int to a size_t before comparing

File: src/libtools/akimaspline.cc [hunk 3]
```diff
@@ -141,7 +136,7 @@ void AkimaSpline::Interpolate(ub::vector<double> &x, ub::vector<double> &y)
     }
 }
 
-void AkimaSpline::Fit(ub::vector<double> &x, ub::vector<double> &y)
+void AkimaSpline::Fit(Eigen::VectorXd &x, Eigen::VectorXd &y)
```
Comment (JoshuaSBrown) [inline src/libtools/akimaspline.cc]: Why do we even have this function?

File: src/libtools/linalg.cc [hunk 4]
```diff
@@ -19,223 +19,142 @@
 #include <iostream>
 #include <sstream>
 namespace votca { namespace tools {
-
-using namespace std;
-namespace ub = boost::numeric::ublas;
```
Comment (JoshuaSBrown) [inline src/libtools/linalg.cc]: This is a .cc source file why did you remove using namespace std?

File: src/libtools/linalg.cc [hunk 5]
```diff
@@ -19,223 +19,142 @@
...
-
-         _diagS(_i,_i) = sqrt(S_eigenvalues[_i]);
-     }
-    ub::matrix<double> _temp = ub::prod( _diagS, ub::trans(S));
-    S = ub::prod( S,_temp );
-    
-  
-    return 1;
-} 
+    bool linalg_eigenvalues(Eigen::MatrixXd&A, Eigen::VectorXd &E, Eigen::MatrixXd&V , int nmax ){
```
Comment (JoshuaSBrown) [inline src/libtools/linalg.cc]: Can you make nmax const?

File: include/votca/tools/linalg.h [hunk 6]
```diff
@@ -17,238 +17,35 @@
...
+    void linalg_constrained_qrsolve(Eigen::VectorXd &x, Eigen::MatrixXd &A,
+            const Eigen::VectorXd &b, const Eigen::MatrixXd &constr);
+  
+  /**
      * \brief solves A*x=b under the constraint B*x = 0
      * @param x storage for x
      * @param A matrix for linear equation system
      * @param b inhomogenity
-     * @param B constrained condition
+     * @param constr constrained condition B (or is it the transposed one? check that)
```
Comment (JoshuaSBrown) [inline include/votca/tools/linalg.h]: This qrsolver is obviously an important function is there anyway we can also provide a link to some documentation that goes into a little more detail about how it works?

File: src/libtools/table.cc [hunk 7]
```diff
@@ -30,12 +30,12 @@ namespace tools {
 using namespace boost;
 using namespace std;
 
-void Table::resize(int N, bool preserve) {
-  _x.resize(N, preserve);
-  _y.resize(N, preserve);
-  _flags.resize(N, preserve);
+void Table::resize(int N) {
```
Comment (JoshuaSBrown) [inline src/libtools/table.cc]: Make int const

File: src/libtools/akimaspline.cc (hunk 3 above)
Comment (JensWehner) [reply src/libtools/akimaspline.cc]: akima must implement this because it is implementing a virtual class

File: src/libtools/linalg.cc (hunk 4 above)
Comment (JensWehner) [reply src/libtools/linalg.cc]: do not need it anymore

File: src/libtools/linalg.cc (hunk 5 above)
Comment (JensWehner) [reply src/libtools/linalg.cc]: it is a copy not a reference, so it does not matter

File: include/votca/tools/linalg.h (hunk 6 above)
Comment (JensWehner) [reply include/votca/tools/linalg.h]: agreed I did nonly reverse engineer the Eigen implementation and I have no real clue how it works.

File: src/libtools/table.cc (hunk 7 above)
Comment (JensWehner) [reply src/libtools/table.cc]: it is a copy of the parameter, making it const serves no purpose

File: include/votca/tools/linalg.h [hunk 8]
```diff
@@ -17,238 +17,35 @@
...
-     * @param returns smallest eigenvalue of S
-     * @param Trace of the product of two matrices
-    
-     * returns the the Trace of the product of two matrices
-     */
-   double linalg_traceofProd(const ub::matrix<double>& A,const ub::matrix<double>& B );
-   
-    /**
-    * \brief solves A*x=b
-    * @param A, first matrix
```
[the remark this answers is not in the document]
Comment (JensWehner) [reply include/votca/tools/linalg.h]: yep

File: include/votca/tools/linalg.h [hunk 9]
```diff
@@ -17,238 +17,35 @@
...
-    * * @param b,  inhomogenity, destroyed and contains the x afterwards
-    
-    
-     * returns the the solves A*x=b for a matrix of b.
-     */
-   bool linalg_solve(const ub::matrix<double> &A, ub::vector<double> &b);
+     bool linalg_eigenvalues(Eigen::MatrixXd&A, Eigen::VectorXd &E, Eigen::MatrixXd&V , int nmax );
+     
+    bool linalg_eigenvalues(Eigen::MatrixXf&A, Eigen::VectorXf &E, Eigen::MatrixXf&V , int nmax );
```
[the remark this answers is not in the document]
Comment (JensWehner) [reply include/votca/tools/linalg.h]: okay, good

File: include/votca/tools/application.h [hunk 10]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright 2009-2011 The VOTCA Development Team (http://www.votca.org)
+ * Copyright 2009-2018 The VOTCA Development Team (http://www.votca.org)
```
Comment (junghans) [inline include/votca/tools/application.h]: Huh, why the update? You didn't touch that file!

Comment (junghans) [reply include/votca/tools/application.h]: Just leave it, it won't hurt. Next time use something like:
```
git diff --name-only origin/master | xargs sed -i '/Copyright/s/-20../-2018/'
```

Comment (JoshuaSBrown) [reply include/votca/tools/application.h]: Is there anyway we can automate this, whenever a push is made. So the files that were changed reflect the dates correctly?

Comment (junghans) [reply include/votca/tools/application.h]: You could make the above command a git pre-commit hook.
